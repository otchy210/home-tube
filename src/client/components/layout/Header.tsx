import clsx from 'clsx';
import { makeStyles } from '@material-ui/core/styles';
import { AppBar, IconButton, Toolbar, Typography } from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import { SIDE_MENU_WIDTH } from './Base';

const useStyles = makeStyles((theme) => ({
    appBar: {
        zIndex: theme.zIndex.drawer + 1,
        transition: theme.transitions.create(['width', 'margin'], {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen,
        }),
    },
    appBarShift: {
        marginLeft: SIDE_MENU_WIDTH,
        width: `calc(100% - ${SIDE_MENU_WIDTH}px)`,
        transition: theme.transitions.create(['width', 'margin'], {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.enteringScreen,
        }),
    },
    toolbar: {
        paddingRight: 24, // keep right padding when drawer closed
    },
    menuButton: {
        marginRight: 36,
    },
    menuButtonHidden: {
        display: 'none',
    },
}));

interface Props {
    menuOpened: boolean;
    openMenu: () => void;
}

const Header: React.FC<Props> = ({ menuOpened, openMenu }) => {
    const classes = useStyles();
    return (
        <AppBar position="absolute" className={clsx(classes.appBar, menuOpened && classes.appBarShift)}>
            <Toolbar className={classes.toolbar}>
                <IconButton
                    edge="start"
                    color="inherit"
                    onClick={openMenu}
                    className={clsx(classes.menuButton, menuOpened && classes.menuButtonHidden)}
                >
                    <MenuIcon />
                </IconButton>
                <Typography>HomeTube</Typography>
            </Toolbar>
        </AppBar>
    );
};

export default Header;
