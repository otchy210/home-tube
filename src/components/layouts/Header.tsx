import React, { useEffect, useRef } from 'react';
import { LinkContainer } from 'react-router-bootstrap';
import styled from 'styled-components';
import Config from '../../images/config.svg';
import Icon from '../../images/icon.svg';
import Language from '../../images/language.svg';
import Logo from '../../images/logo.svg';
import Search from '../../images/search.svg';
import { partiallyPreventDefault } from '../../utils/EventUtils';
import { PrimaryButton } from '../common/buttons';
import { Form, FormSearchInput } from '../common/form';
import { FluidContainer } from '../common/layouts';
import { NavItem, NavLink, NavItemDropdown, NavLinkDropdown, DropdownMenu, DropdownItem } from '../common/nav';
import { Navbar, NavbarBrand, NavbarCollapse, NavbarNav, NavbarToggler } from '../common/navbar';
import { useHomePageQuery } from '../providers/HomePageQueryProvider';
import { LANGUAGES, useI18n } from '../providers/I18nProvider';
import { SearchQuery, useSearchQuery } from '../providers/SearchQueryProvider';
import { Shortcut, useShortcut } from '../providers/ShortcutProvider';

const HomeTubeIcon = styled(Icon)`
    width: 32px;
    height: 32px;
`;

const HomeTubeLogo = styled(Logo)`
    height: 32px;
`;

const SearchIcon = styled(Search)`
    height: 22px;
`;

const LanguageIcon = styled(Language)`
    width: 32px;
    height: 32px;
`;

const ConfigIcon = styled(Config)`
    width: 32px;
    height: 32px;
`;

const IconLabel = styled.span.attrs({ className: 'd-inline d-sm-none d-lg-inline ms-1 ms-sm-0 ms-lg-1 align-middle' })`
    color: rgba(0, 0, 0, 0.55);
`;

const setIfExist = (name: string, refs: { [name: string]: React.RefObject<HTMLInputElement> }, setter: (value: string) => void) => {
    const ref = refs[name];
    if (!ref || !ref.current?.value) {
        return;
    }
    setter(ref.current.value);
};

const getSearchQueryFromRefs = (refs: { [name: string]: React.RefObject<HTMLInputElement> }): SearchQuery => {
    const searchQuery = {} as SearchQuery;
    setIfExist('names', refs, (value) => (searchQuery.names = value.split(/\s/)));
    return searchQuery;
};

const Header: React.FC = () => {
    const { registerShortcut, unregisterShortcut } = useShortcut();
    const { langKey, setLangKey, t } = useI18n();
    const namesRef = useRef<HTMLInputElement>(null);
    const { setPage } = useHomePageQuery();
    const { setSearchQuery } = useSearchQuery();
    const doSearch = () => {
        const updatedSearchQuery = getSearchQueryFromRefs({
            names: namesRef,
        });
        setSearchQuery(updatedSearchQuery);
    };
    const onQueryKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
        // cannot avoid using deprecated `e.keyCode` due to https://qiita.com/ledsun/items/31e43a97413dd3c8e38e
        if (e.keyCode === 13) {
            e.preventDefault();
            doSearch();
            return;
        }
    };
    const onSearchSubmit = (e: React.BaseSyntheticEvent) => {
        e.preventDefault();
        doSearch();
    };
    useEffect(() => {
        const focusSearchInput: Shortcut = {
            keyCode: 191,
            fn: () => {
                namesRef.current?.focus();
            },
        };
        registerShortcut(focusSearchInput);
        return () => {
            unregisterShortcut(focusSearchInput);
        };
    });
    return (
        <Navbar>
            <FluidContainer>
                <NavbarBrand className="me-auto" href="/" onClick={partiallyPreventDefault(() => setPage('1'))}>
                    <HomeTubeIcon className="d-inline d-sm-none" />
                    <HomeTubeLogo className="d-none d-sm-inline" />
                </NavbarBrand>
                <NavbarToggler />
                <NavbarCollapse className="justify-content-end mt-3 mt-sm-0">
                    <NavbarNav>
                        <NavItem>
                            <Form className="d-flex" onSubmit={onSearchSubmit}>
                                <FormSearchInput onKeyDown={onQueryKeyDown} ref={namesRef} />
                                <PrimaryButton className="ms-2" onClick={onSearchSubmit}>
                                    <SearchIcon />
                                    <span className="d-inline d-sm-none d-md-inline ms-1 ms-sm-0 ms-lg-1 align-middle">{t('Search')}</span>
                                </PrimaryButton>
                            </Form>
                        </NavItem>
                        <NavItemDropdown className="ms-2 mt-2 mt-sm-0">
                            <NavLinkDropdown className="p-1">
                                <LanguageIcon />
                                <IconLabel>{t('Language')}</IconLabel>
                            </NavLinkDropdown>
                            <DropdownMenu position="end">
                                {LANGUAGES.map(({ key, label }) => {
                                    return (
                                        <DropdownItem key={`lang-${key}`} onClick={() => setLangKey(key)}>
                                            {key === langKey ? 'â–¸ ' : ''}
                                            {t(label)}
                                        </DropdownItem>
                                    );
                                })}
                            </DropdownMenu>
                        </NavItemDropdown>
                        <NavItem className="ms-2 mt-2 mt-sm-0">
                            <LinkContainer to="/config">
                                <NavLink className="p-1">
                                    <ConfigIcon />
                                    <IconLabel>{t('Config')}</IconLabel>
                                </NavLink>
                            </LinkContainer>
                        </NavItem>
                    </NavbarNav>
                </NavbarCollapse>
            </FluidContainer>
        </Navbar>
    );
};

export default Header;
