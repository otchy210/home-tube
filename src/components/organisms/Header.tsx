import React from 'react';
import { Button, Container, Form, FormControl, Nav, Navbar } from 'react-bootstrap';
import Icon from '../../images/icon.svg';
import Logo from '../../images/logo.svg';
import Config from '../../images/config.svg';

const iconStyle = { height: '32px' };

const Header: React.FC = () => {
    return (
        <Navbar bg="light" variant="light" className="border-bottom" fixed="top" expand="sm">
            <Container fluid>
                <Navbar.Brand href="#" className="me-auto">
                    <Icon className="d-inline d-md-none" style={iconStyle} />
                    <Logo className="d-none d-md-inline" style={iconStyle} />
                </Navbar.Brand>
                <Navbar.Toggle />
                <Navbar.Collapse className="justify-content-end mt-3 mt-sm-0">
                    <Nav>
                        <Form className="d-flex">
                            <FormControl type="search" />
                            <Button variant="primary" className="ms-2">
                                Search
                            </Button>
                        </Form>
                    </Nav>
                    <Nav>
                        <Nav.Link href="#features" className="ms-1">
                            <Config style={iconStyle} />
                            <span className="d-inline d-sm-none d-lg-inline ms-2 ms-sm-0 ms-lg-2 align-middle">Config</span>
                        </Nav.Link>
                    </Nav>
                </Navbar.Collapse>
            </Container>
        </Navbar>
    );
};

export default Header;
