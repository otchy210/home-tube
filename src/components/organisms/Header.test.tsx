import React from 'react';
import { render } from '@testing-library/react';
import Header from './Header';
import { BrowserRouter } from 'react-router-dom';

describe('Header', () => {
    it('renders properly', () => {
        const { container } = render(
            <BrowserRouter>
                <Header />
            </BrowserRouter>
        );
        expect(container).toMatchSnapshot();
    });
});
