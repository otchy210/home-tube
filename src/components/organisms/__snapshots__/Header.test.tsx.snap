// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders properly 1`] = `
<div>
  <nav
    class="border-bottom navbar navbar-expand-sm navbar-light bg-light fixed-top"
  >
    <div
      class="container-fluid"
    >
      <a
        class="me-auto navbar-brand"
        href="/"
      >
        [MockedIconSvg className="sc-bdvvtL jLKyDm d-inline d-sm-none" /]
        [MockedLogoSvg className="sc-gsDKAQ ljCoMj d-none d-sm-inline" /]
      </a>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end mt-3 mt-sm-0 navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <form
            class="d-flex"
          >
            <input
              class="form-control"
              type="search"
            />
            <button
              class="ms-2 text-nowrap btn btn-primary"
              type="button"
            >
              [MockedSearchSvg className="sc-dkPtRN gtPUWC" /]
              <span
                class="d-inline d-sm-none d-md-inline ms-1 ms-sm-0 ms-lg-1 align-middle"
              >
                Search
              </span>
            </button>
          </form>
        </div>
        <div
          class="ms-2 mt-2 mt-sm-0 navbar-nav"
        >
          <div>
            [MockedLanguageSvg className="sc-hKwDye jjkbGY" /]
            <span
              class="sc-jRQBWg fmpVRd d-inline d-sm-none d-lg-inline ms-1 ms-sm-0 ms-lg-1 align-middle"
            >
              Language
            </span>
          </div>
        </div>
        <div
          class="ms-2 mt-2 mt-sm-0 navbar-nav"
        >
          <a
            class="p-0 nav-link"
            data-rr-ui-event-key="/config"
            href="/config"
          >
            [MockedConfigSvg className="sc-eCImPb gpLiHk" /]
            <span
              class="sc-jRQBWg fmpVRd d-inline d-sm-none d-lg-inline ms-1 ms-sm-0 ms-lg-1 align-middle"
            >
              Config
            </span>
          </a>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
